shared/fonts/fonts.cpp

```diff
@@ -0,0 +1,295 @@
+
+   init_font_search_dirs();
+
+    return (error == 0);
+  }
+
+  static const FT_Library& GetLibrary() {
+    return library;
+  }
+
+  ~FontManager() {
+    FT_Done_FreeType(library);
+  }
+
+private:
+  FontManager() = default;
+  static FT_Library library;
+};
+
+FT_Library FontManager::library;
+static bool FreeTypeAlive = FontManager::Init();
+
+bool font_load(std::string fName, unsigned size, RawFont& rawFont) {
+
+  if (!enigma::fonts::FreeTypeAlive)
+    return false;
+
+  FT_Face face;
+  FT_GlyphSlot slot;
+  FT_Error error;
+
+  error = FT_New_Face(enigma::fonts::FontManager::GetLibrary(), fName.c_str(), 0, &face );
+
+  if (error != 0)
+    return false;
+
+  FT_SfntName name;
+  FT_Get_Sfnt_Name(face, TT_NAME_ID_FULL_NAME, &name);
+  rawFont.name = std::string(reinterpret_cast<char*>(name.string), name.string_len);
+
+  // GameMaker has always generated fonts at 96 dpi, default Windows dpi since 1980s
+  error = FT_Set_Char_Size( face, size * 64, 0, 96, 0); // 96 dpi, 64 is 26.6 fixed point conversion
+
+  if (error != 0)
+    return false;
+
+  slot = face->glyph;
+
+  // sum of all glyphs in all ranges
+  size_t gcount = 0;
+  for (const GlyphRange& range : rawFont.ranges) gcount += range.glyphs.size();
+ 
+  // vector of all (raw) glyphs used later for sorting
+  rawFont.glyphs.resize(gcount);
+   
+  // loop over all ranges
+  size_t currOffset = 0;
+  for (GlyphRange& range : rawFont.ranges) {
+    // loop over each character per range
+    for (size_t c = 0; c < range.size(); ++c) {
+      // open character to get metrics and pixel data
+      error = FT_Load_Char(face, range.start + c, FT_LOAD_RENDER);
+
+      #if defined(DEBUG_MODE) || !defined(ENIGMA_GAME)
+      if (error != 0)
+        DEBUG_MESSAGE("Freetype error loading char: " + std::to_string(range.start + c), MESSAGE_TYPE::M_ERROR);
+      #endif
+      
+       FT_Bitmap& charBitmap = slot->bitmap;
+       FT_Glyph_Metrics fftMetrics = face->glyph->metrics;
+
+       RawGlyph rg;
+       rg.character = range.start + c;
+       rg.dimensions = pvrect(0, 0, charBitmap.width, charBitmap.rows, -1);
+       rg.range = &range;
+       rg.horiBearingX = fftMetrics.horiBearingX / 64;
+       rg.horiBearingY = fftMetrics.horiBearingY / 64;
+       rg.linearHoriAdvance = face->glyph->linearHoriAdvance / 65536; // more fixed point conversion
+
+       if (charBitmap.width > 0 && charBitmap.rows > 0) {
+         rg.pxdata = new unsigned char[charBitmap.width * charBitmap.rows];
+       
+         // copy glyph to texture
+         for (unsigned y = 0; y < charBitmap.rows; ++y) {
+           for (unsigned x = 0; x < charBitmap.width; ++x) {
+             unsigned index = (y * charBitmap.pitch + x);
+             rg.pxdata[index] = charBitmap.buffer[index];
+            }
+          }
+       }
+       
+       rawFont.glyphs[currOffset++] = rg;
+    }
+  }
+
+  // FIXME: idk how GM calculates the line height 
```

Comment:
IDK what to do here. I think GM does something like using max glyph height but freetype recommends a different value for line height.

Reply:
The GML function `string_height` implementation is probably the answer. We have that function in the engine so maybe look at it, I'm sure the frontend would be written to be consistent with the engine right? Yeah.

Reply:
pretty sure that will just return the line height it does here